btree/btree.c

```diff
@@ -2034,12 +2151,628 @@ static void generic_move(struct nd *src, struct nd *tgt,
+		       m0_vec_count(&new_rec.r_key.k_data.ov_vec));
+	m0_bufvec_copy(&node_slot->s_rec.r_val, &new_rec.r_val,
+		       m0_vec_count(&new_rec.r_val.ov_vec));
+	node_done(node_slot, tx, true);
+}
+
+#endif
+
+static void alloc(struct m0_btree_oimpl *oi, int height)
+{
+	oi->i_level = m0_alloc(height * sizeof(struct level));
+}
+
+# if 0
+static int callback_temp(struct m0_btree_rec *rec, int64_t key)
+{
+	struct m0_btree_rec temp_rec;
+	int64_t val = key + 1;
+	void* p_key;
+	void *p_val;
+
+	m0_bcount_t ksize = MAX_KEY_SIZE ;
+	m0_bcount_t vsize = MAX_VAL_SIZE ;
+	p_key = &key;
+	p_val = &val;
+
+	temp_rec.r_key.k_data.ov_vec.v_nr     = 1;
+	temp_rec.r_key.k_data.ov_vec.v_count  = &ksize;
+	temp_rec.r_key.k_data.ov_buf          = &p_key;
+
+	temp_rec.r_val.ov_vec.v_nr           = 1;
+	temp_rec.r_val.ov_vec.v_count        = &vsize;
+	temp_rec.r_val.ov_buf                = &p_val;
+
+	m0_bufvec_copy(&rec->r_key.k_data,
+		       &temp_rec.r_key.k_data,
+		       m0_vec_count(&temp_rec.r_key.k_data.ov_vec));
+	m0_bufvec_copy(&rec->r_val, &temp_rec.r_val,
+		       m0_vec_count(&temp_rec.r_val.ov_vec));
+	return 0;
+}
+#endif
+/*get_tick for insert operation
+static int64_t get_tick_insert(struct m0_btree_op *bop)*/
+static int64_t get_tick_insert(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop   = M0_AMB(bop, smop, bo_op);
+	struct td             *tree  = bop->bo_arbor->t_tree;
+	uint64_t               flags = bop->bo_flags;
+	struct m0_btree_oimpl *oi    = bop->bo_i;
+	struct level          *level = &oi->i_level[oi->i_used];
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((flags & OF_COOKIE) && cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:
+		if (cookie_is_valid(tree, bop->bo_rec.r_key) &&
+		    !isOverflowPossible(bop->bo_rec.r_key.k_cookie.node))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	case P_SETUP: {
+		alloc(bop->bo_i, tree->t_height);
+		if (bop->bo_i == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		return P_LOCKALL;
+	}
+	case P_LOCKALL:
+		if (bop->bo_flags & OF_LOCKALL)
+			return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr, P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    node_slot = {};
+			struct segaddr down;
+
+			level->l_node = oi->i_nop.no_node;
+			node_slot.s_node = oi->i_nop.no_node;
+			/*int rc = node_verify(level->l_node);
+			if(rc != 0 )
+			{
+				// return fail
+				printf("rc for node_verify :%d\n", rc);
+			}*/
+			oi->i_nop.no_node = NULL;
+			node_op_fini(&oi->i_nop);
+			node_find(&node_slot, &bop->bo_rec.r_key);
+			level->l_idx = node_slot.s_idx;
+			if (node_level(node_slot.s_node) > 0) {
+				//validate node_slot.s_node(i.e. is it still exists or not)
+				if (!is_node_valid(node_slot.s_node))
```

Comment:
Why are we validating after we have already referenced this node in node_find? You want to validate before trusting the node contents.